File: frontend/src/services/cloud.ts
import api, { PageData } from './api'

export interface CloudAccount {
  id: string
  name: string
  provider: string
  region: string
  status: number
  description: string
  created_by: string
  created_at: string
}

export interface CloudInstance {
  id: string
  account_id: string
  account?: CloudAccount
  instance_id: string
  name: string
  provider: string
  region: string
  zone: string
  instance_type: string
  image_id: string
  cpu: number
  memory: number
  public_ip: string
  private_ip: string
  status: string
  host_id?: string
  host?: { id: string; name: string; ip: string }
  cluster_id?: string
  node_pool_id?: string
  expire_at?: string
  charge_type: string
  created_at: string
}

export interface NodePool {
  id: string
  cluster_id: string
  cluster?: { id: string; name: string; code: string }
  account_id: string
  account?: CloudAccount
  name: string
  instance_type: string
  image_id: string
  min_size: number
  max_size: number
  desired_size: number
  current_size: number
  status: string
  description: string
  created_at: string
}

export interface Region { id: string; name: string }
export interface Zone { id: string; name: string; region_id: string }
export interface InstanceTypeInfo { id: string; name: string; cpu: number; memory: number }
export interface Image { id: string; name: string; os: string }

export const cloudService = {
  // 云账号
  listAccounts: (params: { page?: number; page_size?: number; provider?: string; keyword?: string }) =>
    api.get<unknown, { data: PageData<CloudAccount> }>('/cloud/accounts', { params }),
  createAccount: (data: { name: string; provider: string; access_key: string; secret_key: string; region?: string; description?: string }) =>
    api.post<unknown, { data: CloudAccount }>('/cloud/accounts', data),
  getAccount: (id: string) =>
    api.get<unknown, { data: CloudAccount }>(`/cloud/accounts/${id}`),
  updateAccount: (id: string, data: Partial<{ name: string; access_key: string; secret_key: string; region: string; description: string }>) =>
    api.put<unknown, { data: CloudAccount }>(`/cloud/accounts/${id}`, data),
  deleteAccount: (id: string) =>
    api.delete(`/cloud/accounts/${id}`),
  verifyAccount: (id: string) =>
    api.post(`/cloud/accounts/${id}/verify`),

  // 云资源查询
  listRegions: (accountId: string) =>
    api.get<unknown, { data: Region[] }>(`/cloud/accounts/${accountId}/regions`),
  listZones: (accountId: string, region: string) =>
    api.get<unknown, { data: Zone[] }>(`/cloud/accounts/${accountId}/zones`, { params: { region } }),
  listInstanceTypes: (accountId: string, region: string) =>
    api.get<unknown, { data: InstanceTypeInfo[] }>(`/cloud/accounts/${accountId}/instance-types`, { params: { region } }),
  listImages: (accountId: string, region: string) =>
    api.get<unknown, { data: Image[] }>(`/cloud/accounts/${accountId}/images`, { params: { region } }),

  // 云实例
  listInstances: (params: { page?: number; page_size?: number; account_id?: string; status?: string; cluster_id?: string }) =>
    api.get<unknown, { data: PageData<CloudInstance> }>('/cloud/instances', { params }),
  createInstance: (data: { account_id: string; name: string; region: string; zone?: string; instance_type: string; image_id: string; charge_type?: string }) =>
    api.post<unknown, { data: CloudInstance }>('/cloud/instances', data),
  getInstance: (id: string) =>
    api.get<unknown, { data: CloudInstance }>(`/cloud/instances/${id}`),
  startInstance: (id: string) =>
    api.post<unknown, { data: CloudInstance }>(`/cloud/instances/${id}/start`),
  stopInstance: (id: string) =>
    api.post<unknown, { data: CloudInstance }>(`/cloud/instances/${id}/stop`),
  terminateInstance: (id: string) =>
    api.post<unknown, { data: CloudInstance }>(`/cloud/instances/${id}/terminate`),
  syncInstance: (id: string) =>
    api.post<unknown, { data: CloudInstance }>(`/cloud/instances/${id}/sync`),
  bindHost: (id: string) =>
    api.post<unknown, { data: CloudInstance }>(`/cloud/instances/${id}/bindhost`),
  unbindHost: (id: string) =>
    api.post<unknown, { data: CloudInstance }>(`/cloud/instances/${id}/unbindhost`),

  // 节点池
  listNodePools: (params: { page?: number; page_size?: number; cluster_id?: string }) =>
    api.get<unknown, { data: PageData<NodePool> }>('/cloud/nodepools', { params }),
  createNodePool: (data: { cluster_id: string; account_id: string; name: string; instance_type?: string; image_id?: string; min_size?: number; max_size?: number; desired_size?: number; description?: string }) =>
    api.post<unknown, { data: NodePool }>('/cloud/nodepools', data),
  getNodePool: (id: string) =>
    api.get<unknown, { data: NodePool }>(`/cloud/nodepools/${id}`),
  updateNodePool: (id: string, data: Partial<{ name: string; min_size: number; max_size: number; desired_size: number; description: string }>) =>
    api.put<unknown, { data: NodePool }>(`/cloud/nodepools/${id}`, data),
  deleteNodePool: (id: string) =>
    api.delete(`/cloud/nodepools/${id}`),
  scaleNodePool: (id: string, desired_size: number) =>
    api.post<unknown, { data: NodePool }>(`/cloud/nodepools/${id}/scale`, { desired_size }),
}

File: backend/internal/model/cloud.go
package model

import (
	"time"

	"github.com/google/uuid"
	"gorm.io/gorm"
)

// CloudAccount 云账号
type CloudAccount struct {
	ID          uuid.UUID      `json:"id" gorm:"type:uuid;primary_key"`
	Name        string         `json:"name" gorm:"size:100;not null"`
	Provider    string         `json:"provider" gorm:"size:20;not null;index"` // alicloud, aws
	AccessKey   string         `json:"-" gorm:"size:500;not null"`
	SecretKey   string         `json:"-" gorm:"size:500;not null"`
	Region      string         `json:"region" gorm:"size:50"`
	Status      int            `json:"status" gorm:"default:1;index"` // 1=normal, 0=disabled, 2=verify_failed
	Description string         `json:"description" gorm:"size:255"`
	CreatedBy   uuid.UUID      `json:"created_by" gorm:"type:uuid"`
	CreatedAt   time.Time      `json:"created_at"`
	UpdatedAt   time.Time      `json:"updated_at"`
	DeletedAt   gorm.DeletedAt `json:"-" gorm:"index"`
}

func (a *CloudAccount) BeforeCreate(tx *gorm.DB) error {
	if a.ID == uuid.Nil {
		a.ID = uuid.New()
	}
	return nil
}

// CloudInstance 云主机实例
type CloudInstance struct {
	ID           uuid.UUID      `json:"id" gorm:"type:uuid;primary_key"`
	AccountID    uuid.UUID      `json:"account_id" gorm:"type:uuid;index;not null"`
	Account      *CloudAccount  `json:"account,omitempty" gorm:"foreignKey:AccountID"`
	InstanceID   string         `json:"instance_id" gorm:"size:100;index"`
	Name         string         `json:"name" gorm:"size:100;not null"`
	Provider     string         `json:"provider" gorm:"size:20;not null;index"`
	Region       string         `json:"region" gorm:"size:50"`
	Zone         string         `json:"zone" gorm:"size:50"`
	InstanceType string         `json:"instance_type" gorm:"size:50"`
	ImageID      string         `json:"image_id" gorm:"size:100"`
	CPU          int            `json:"cpu" gorm:"default:0"`
	Memory       int            `json:"memory" gorm:"default:0"` // MB
	PublicIP     string         `json:"public_ip" gorm:"size:50"`
	PrivateIP    string         `json:"private_ip" gorm:"size:50"`
	Status       string         `json:"status" gorm:"size:20;default:'pending';index"` // pending, running, stopped, terminated
	HostID       *uuid.UUID     `json:"host_id" gorm:"type:uuid;index"`
	Host         *Host          `json:"host,omitempty" gorm:"foreignKey:HostID"`
	ClusterID    *uuid.UUID     `json:"cluster_id" gorm:"type:uuid;index"`
	NodePoolID   *uuid.UUID     `json:"node_pool_id" gorm:"type:uuid;index"`
	ExpireAt     *time.Time     `json:"expire_at"`
	ChargeType   string         `json:"charge_type" gorm:"size:20"` // PostPaid, PrePaid
	CreatedBy    uuid.UUID      `json:"created_by" gorm:"type:uuid"`
	CreatedAt    time.Time      `json:"created_at" gorm:"index"`
	UpdatedAt    time.Time      `json:"updated_at"`
	DeletedAt    gorm.DeletedAt `json:"-" gorm:"index"`
}

func (i *CloudInstance) BeforeCreate(tx *gorm.DB) error {
	if i.ID == uuid.Nil {
		i.ID = uuid.New()
	}
	return nil
}

// NodePool K8s节点池
type NodePool struct {
	ID           uuid.UUID      `json:"id" gorm:"type:uuid;primary_key"`
	ClusterID    uuid.UUID      `json:"cluster_id" gorm:"type:uuid;index;not null"`
	Cluster      *Cluster       `json:"cluster,omitempty" gorm:"foreignKey:ClusterID"`
	AccountID    uuid.UUID      `json:"account_id" gorm:"type:uuid;index;not null"`
	Account      *CloudAccount  `json:"account,omitempty" gorm:"foreignKey:AccountID"`
	Name         string         `json:"name" gorm:"size:100;not null"`
	InstanceType string         `json:"instance_type" gorm:"size:50"`
	ImageID      string         `json:"image_id" gorm:"size:100"`
	MinSize      int            `json:"min_size" gorm:"default:0"`
	MaxSize      int            `json:"max_size" gorm:"default:10"`
	DesiredSize  int            `json:"desired_size" gorm:"default:0"`
	CurrentSize  int            `json:"current_size" gorm:"default:0"`
	Status       string         `json:"status" gorm:"size:20;default:'active'"` // active, scaling, error
	Description  string         `json:"description" gorm:"size:255"`
	CreatedBy    uuid.UUID      `json:"created_by" gorm:"type:uuid"`
	CreatedAt    time.Time      `json:"created_at"`
	UpdatedAt    time.Time      `json:"updated_at"`
	DeletedAt    gorm.DeletedAt `json:"-" gorm:"index"`
}

func (n *NodePool) BeforeCreate(tx *gorm.DB) error {
	if n.ID == uuid.Nil {
		n.ID = uuid.New()
	}
	return nil
}
